src/serialize.h

```diff
@@ -446,6 +428,16 @@ class CFlatData
         s.read(pbegin, pend - pbegin);
     }
 };
+//! Construct a FlatRange wrapper around a const vector.
+template<typename T> static inline const FlatRangeWrapper<const char> FlatVector(const T& t) { return FlatRangeWrapper<const char>((const char*)t.data(), (const char*)(t.data() + t.size())); }
+//! Construct a FlatRange wrapper around a non-const vector.
+template<typename T> static inline FlatRangeWrapper<char> FlatVector(T& t) { return FlatRangeWrapper<char>((char*)t.data(), (char*)(t.data() + t.size())); }
+//! Construct a FlatRange wrapper around a const POD and array types.
+template<typename T> static inline const FlatRangeWrapper<const char> FlatDataInner(const T* t, size_t len) { return FlatRangeWrapper<const char>((const char*)t, ((const char*)t) + len); }
+//! Construct a FlatRange wrapper around a non-const POD and array types.
+template<typename T> static inline FlatRangeWrapper<char> FlatDataInner(T* t, size_t len) { return FlatRangeWrapper<char>((char*)t, ((char*)t) + len); }
+//! Helper macro to easily serialize POD types.
+#define FLATDATA(x) FlatDataInner(&(x), sizeof(x))
```

Comment:
In commit "Generalize FlatData wrapper"

Though these changes don't make FLATDATA more dangerous than it was previously, the lack of type safety relying on C casts and sizeof here is a little scary. I experimented a little, and it seems this could be cleaned up with some simple changes I posted here: 6ef78bcd83dd6f88362dec29736811b762ad75eb. Could you take a look, and maybe incorporate these into the PR?

Reply:
I've incorporated part of your changes, but gone further and just added native support for serializing char arrays (without any wrapper). I think this is much cleaner now.